### ISSUE TITLE
nil pointer dereference when inserting data from a []interface{}

### ISSUE THREAD

**Issue by deleted** [jackc/pgx#1348]
**Describe the bug**
When inserting slice data to a postgres ARRAY column (or copying with `CopyTo`), if the type of the input data is `[]interface{}`, pgx tries to dereference a `nil` pointer and segfaults.

**To Reproduce**

```go
package main

import (
	"context"
        "os"

	"github.com/jackc/pgx/v5"
)

func main() {
	ctx := context.Background()

	DB_URL := os.GetEnv("DATABASE_URL")
	panicOnErr := func(err error) {
		if err != nil {
			panic(err)
		}
	}

	conn, err := pgx.Connect(ctx, DB_URL)
	panicOnErr(err)
	defer conn.Close(ctx)

	_, err = conn.Exec(ctx, `
	  CREATE TEMPORARY TABLE foo (
		stuff int[]
	  )
	`)
	panicOnErr(err)

	qry := `
	  INSERT INTO foo (stuff) values ($1)
	`
	concreteRows := []int{1, 2, 3, 4, 5}
	_, err = conn.Exec(ctx, qry, concreteRows) // Works fine
	panicOnErr(err)

	interfaceRows := []interface{}{1, 2, 3, 4, 5} // SIGSEGV!
	_, err = conn.Exec(ctx, qry, interfaceRows)
	panicOnErr(err)
}
```

**Expected behavior**
Ideally pgx would either be able to convert the data to concrete values and insert it or else return some sort of error.

**Actual behavior**
The runtime panics with a SIGSEGV, as follows:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x98 pc=0x68d60d]

goroutine 1 [running]:
github.com/jackc/pgx/v5/pgtype.TryWrapDerefPointerEncodePlan({0x0?, 0x0?})
        /home/ted/go/pkg/mod/github.com/jackc/pgx/v5@v5.0.2/pgtype/pgtype.go:1451 +0x4d
github.com/jackc/pgx/v5/pgtype.(*Map).PlanEncode(0xc00012c540, 0x17, 0x1, {0x0?, 0x0})
        /home/ted/go/pkg/mod/github.com/jackc/pgx/v5@v5.0.2/pgtype/pgtype.go:1325 +0x2e6
github.com/jackc/pgx/v5/pgtype.(*ArrayCodec).PlanEncode(0xc00012a158, 0xc00012c540, 0x3ef, 0x1, {0x833200?, 0xc000012180?})
        /home/ted/go/pkg/mod/github.com/jackc/pgx/v5@v5.0.2/pgtype/array_codec.go:61 +0x7b
github.com/jackc/pgx/v5/pgtype.(*Map).PlanEncode(0xc00012c540, 0x3ef, 0x1, {0x833200?, 0xc000012180})
        /home/ted/go/pkg/mod/github.com/jackc/pgx/v5@v5.0.2/pgtype/pgtype.go:1315 +0x1ee
github.com/jackc/pgx/v5/pgtype.(*Map).PlanEncode(0xc00012c540, 0x3ef, 0x1, {0x7ce9e0?, 0xc000012168})
        /home/ted/go/pkg/mod/github.com/jackc/pgx/v5@v5.0.2/pgtype/pgtype.go:1326 +0x314
github.com/jackc/pgx/v5/pgtype.(*Map).Encode(0x7ce9e0?, 0x12168?, 0xc0?, {0x7ce9e0, 0xc000012168}, {0xc000096600, 0x0, 0x80})
        /home/ted/go/pkg/mod/github.com/jackc/pgx/v5@v5.0.2/pgtype/pgtype.go:1969 +0x5f
github.com/jackc/pgx/v5.(*ExtendedQueryBuilder).encodeExtendedParamValue(0xc00014dce0, 0xc00014deb8?, 0x2?, 0x0?, {0x7ce9e0, 0xc000012168})
        /home/ted/go/pkg/mod/github.com/jackc/pgx/v5@v5.0.2/extended_query_builder.go:125 +0xdd
github.com/jackc/pgx/v5.(*ExtendedQueryBuilder).appendParam(0xc00014dce0, 0xc00011cfc0?, 0x3ef?, 0x1, {0x7ce9e0?, 0xc000012168?})
        /home/ted/go/pkg/mod/github.com/jackc/pgx/v5@v5.0.2/extended_query_builder.go:75 +0x16a
github.com/jackc/pgx/v5.(*ExtendedQueryBuilder).appendParam(0xc00014dce0, 0xc00012c540, 0x100c01?, 0xc0?, {0x7ce9e0?, 0xc000012168})
        /home/ted/go/pkg/mod/github.com/jackc/pgx/v5@v5.0.2/extended_query_builder.go:55 +0xe5
github.com/jackc/pgx/v5.(*ExtendedQueryBuilder).Build(0xc00014dce0, 0xc00012c540, 0xc000092050, {0xc0000240c0?, 0x1, 0x1})
        /home/ted/go/pkg/mod/github.com/jackc/pgx/v5@v5.0.2/extended_query_builder.go:36 +0x409
github.com/jackc/pgx/v5.(*Conn).execPrepared(0xc00014dc20, {0x904570, 0xc000120000}, 0xc000092050, {0xc0000240c0?, 0x0?, 0x29?})
        /home/ted/go/pkg/mod/github.com/jackc/pgx/v5@v5.0.2/conn.go:493 +0x65
github.com/jackc/pgx/v5.(*Conn).exec(0xc00014dc20, {0x904570, 0xc000120000}, {0x874c90, 0x29}, {0xc0000240c0?, 0x0?, 0x7f30fda1a118?})
        /home/ted/go/pkg/mod/github.com/jackc/pgx/v5@v5.0.2/conn.go:436 +0x353
github.com/jackc/pgx/v5.(*Conn).Exec(0xc00014dc20, {0x904570?, 0xc000120000?}, {0x874c90, 0x29}, {0xc0000240c0, 0x1, 0x1})
        /home/ted/go/pkg/mod/github.com/jackc/pgx/v5@v5.0.2/conn.go:382 +0x13a
exit status 2
```

**Version**
 - Go: `$ go version` -> go version go1.19.1 linux/amd64
 - PostgreSQL: `$ psql --no-psqlrc --tuples-only -c 'select version()'` ->  PostgreSQL 14.5 (Ubuntu 14.5-1.pgdg22.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0, 64-bit
 - pgx: `$ grep 'github.com/jackc/pgx/v[0-9]' go.mod` -> github.com/jackc/pgx/v5 v5.0.2

**Additional context**
Obviously in the test case above, this can by solved by putting the data in a slice with a concrete element type, but in the program I'm writing I won't know the element type until runtime.

**Comment by deleted** [comment]
This seems to be related to #1335 .

If I apply the changes from #1336 and also add a similar nil check within `pgtype.TryWrapDerefPointerEncodePlan()` it avoids the SEGFAULT, and I get a more helpful error:

```
panic: failed to encode args[0]: unable to encode []interface {}{1, 2, 3, 4, 5} into binary format for _int4 (OID 1007): cannot find encode plan
```

**Comment by jackc** [comment]
Fixed as of 48b4807b333cbe5a365cfc051493d70dee372b58.

As far as handling `[]any` instead of requiring a typed slice, it might be possible, but I think the complexity would outweigh the benefit.
